### ISSUE TITLE
Domain specific annotation of pending_xrefs made by fields.

### ISSUE THREAD

**Issue by jakobandersen** [sphinx-doc/sphinx#2665]
The C++ domain is annotating each pending_xref with information of which scope the xref should be resolved in. While trying to fix #2607 I couldn't find a way to annotate the pending_xrefs created for fields.
The following example illustrates the problem, both with the C++ and Python domain.

``` rst
.. cpp:class:: NS::MyException : public std::exception

    Type of an exception.

.. cpp:namespace:: NS

.. cpp:function:: void my_function()

    :exception NS\:\:MyException:   This works.
    :exception MyException:         This doesn't.

    * Referencing :cpp:class:`NS::MyException` works.
    * Referencing :cpp:class:`MyException` also works.

.. py:class:: PyClass

    .. py:exception:: Ex

    .. py:method:: f()

        :raises PyClass.Ex: This works.
        :raises Ex: This doesn't.

        * Referencing :py:class:`PyClass.Ex` works.
        * Referencing :py:class:`Ex` doesn't work (a bug?)

.. py:method:: g()

    :raises PyClass.Ex: This works.
    :raises Ex: This doesn't, and shouldn't.
```

Is there a mechanism I didn't notice? Otherwise, my best suggestion is that the field uses the domain name to invoke a callback. Is there a prettier way?

**Comment by tk0miya** [comment]
Agreed. Any callbacks to domains are needed to create pending_xrefs for fields.

BTW, 

```
* Referencing :py:class:`Ex` doesn't work (a bug?)
```

I believe this is a bug of python domain.
In python domain, `py:class` and `py:exception` updates the context::

```
.. py:class:: PyClass

   ref_context['py:class'] is 'PyClass' here

ref_context['py:class'] is None here

.. py:exception:: Ex

   ref_context['py:class'] is 'Ex' here

ref_context['py:class'] is None here
```

This behavior has problem as you said. The context is reset after nested definition. Because they simply override the context.

```
.. py:class:: PyClass

    ref_context['py:class'] is `PyClass` here

    .. py:exception:: Ex

       ref_context['py:class'] is `Ex` here

    ref_context['py:class'] is None here
```

It should be better if the contexts work like a stack.

**Comment by jakobandersen** [comment]
Fixed with PR #3528.
